### PR TITLE
APEXMALHAR-2017 Use pre checkpoint notification to optimize operator IO

### PR DESCRIPTION
**Pull request by pramodin**
@chandnisingh @siyuanh @DT-Priyanka please review as it touches many operators

### CODE REVIEW

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java [hunk 1]
```diff
@@ -1195,6 +1188,24 @@ public void close() throws IOException
...
   @Override
+  public void beforeCheckpoint(long l)
+  {
+    try {
+      Map<String, FSFilterStreamContext> openStreams = streamsCache.asMap();
+      for (FSFilterStreamContext streamContext: openStreams.values()) {
+        long start = System.currentTimeMillis();
+        streamContext.finalizeContext();
+        totalWritingTime += System.currentTimeMillis() - start;
+        //streamContext.resetFilter();
```
Comment (otterc) [inline library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java:1200]: why this commented out code?

**Comment by otterc** [comment]
changes to the file output operator look good to me other than the minor comment.

**Comment by gauravgopi123** [comment]
@PramodSSImmaneni: Say this PR goes in 3.5. Will user be able to use this with apex version < 3.3 as new interface CheckpointNotificationListener is add in apex 3.3?

**Comment by pramodin** [comment]
@gauravgopi123 That would be forward compatibility isn't it (old apex version supporting newer malhar versions). There would almost always be newer features added in newer minor versions and we cannot expect old versions to be forward compatible in that fashion.

**Pull request closed without merging**